File: Arrays/229-Majority-Element-II/Python/229-Majority-Element-II-Better.py
"""
Problem: 229. Majority Element II
Approach: Better - Hash Map

Store the frequency of each element.
Whenever an element's frequency becomes greater than n / 3,
add it to the result.

Time Complexity: O(n)
Space Complexity: O(n)
"""

class Solution:
    def majorityElement(self, nums: list[int]) -> list[int]:
        n = len(nums)
        freq = {}
        result = []

        for num in nums:
            freq[num] = freq.get(num, 0) + 1

            if freq[num] > n / 3 and num not in result:
                result.append(num)

            if len(result) == 2:
                break

        return result

File: Arrays/229-Majority-Element-II/JavaScript/229-Majority-Element-II-Better.js
/**
 * Problem: 169. Majority Element
 * Approach: Better - Hash Map
 *
 * Store the frequency of each element.
 * As soon as an element's frequency becomes greater than n / 2,
 * return it.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function(nums) {
    const freq = new Map();
    const n = nums.length;

    for (const num of nums) {
        freq.set(num, (freq.get(num) || 0) + 1);

        if (freq.get(num) > n / 2) {
            return num;
        }
    }

    return -1;
};